library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +95,65 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        assert!(mem::size_of::<T>() == 0);
+
+        const fn max(a: usize, b: usize) -> usize {
+            if a > b { a } else { b }
+        }
+
+        // We are using `[T; 0]` here instead of `T` because
+        // compiler instantiates this code even for non-ZST,
+        // and reports a pointer going beyond the allocation.
+        let alloc: &[T; 0] = const {
```

Comment:
I don't think that the question of subobject provenance has been resolved yet. I suggest to this a `&<Dyn as Pointee>::Metadata` to the start of the metadata and then do a `pointer::add` to get the pointer at runtime (this will be optimized out).

Reply:
Ah right this should have raw pointer type ideally, not reference type.

Reply:
Changed to `&<Dyn as Pointee>::Metadata`. We cannot use raw pointer, compiler says something about mutable pointers in constants.

Reply:
That's still not right though if subobject provenance is in place. You'd get back a pointer that can only be used for `size_of::<<Dyn as Pointee>::Metadata>()` many bytes from its starting address. Or is that suficient?

Reply:
```
let alloc: &<Dyn as Pointee>::Metadata = const { ... }
```

`T` is zero sized, so pointer to `T` is at `alloc+size_of<<Dyn as Pointee>::Metadata()` so it is within `alloc..alloc+size_of<<Dyn as Pointee>::Metadata()`. So I think pointer is fine?

(I don't know what is "subobject provenance"; I quickly googled, found your issue and explanation in unsafe guideline, and still don't understand it.).

Reply:
Where in the `[ padding | header | value ]` block does alloc` point to? For `&<Dyn as Pointee>::Metadata` to be the correct type, it *must* point to the beginning of the header.

Reply:
`alloc` points to the beginning of the header. Past the padding.

Reply:
Okay, that works then, thanks.